pkg/log/log.go

```diff
@@ -0,0 +1,46 @@
+//
+// Copyright 2022 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+)
+
+type ResourceChangeType string
+
+const (
+	ResourceCreate   ResourceChangeType = "Create"
```

Comment:
How is `Create` / `Update` / `Delete` being defined in the context of audit logging? Is it "Resource xyz was created" on Kubernetes, or "The Controller has finished reconciling resource xyz" for the first time

Reply:
If it's the former, we can define custom Create/Update/Delete funcs that does the event log before calling reconcile:

e.g.

https://github.com/redhat-appstudio/application-service/blob/3c370f5f7c6f0963ba44c1fd57554fb379f9f30e/controllers/applicationsnapshotenvironmentbinding_controller.go#L302-L312

Reply:
CC @kim-tsao

Reply:
@johnmcollier  good point, we don't want to be logging in the reconcile loop.  I think it's reasonable to log based on the former

Reply:
that's a good point! but how would we audit CDQ event then. when the cdq is created, it's not completed. and the cdq creation time is just start processing the detection.

Reply:
So `ResourceChangeType` will probably be different from `ConditionChangeType`, whereby `ConditionChangeType` is a subset of `ResourceChangeType`

```
Conditions:
    Last Transition Time:  2022-10-11T17:23:44Z
    Message:               ComponentDetectionQuery is processing
    Reason:                Success
    Status:                True
    Type:                  Processing
    Last Transition Time:  2022-10-11T17:23:46Z
    Message:               ComponentDetectionQuery has successfully finished
    Reason:                OK
    Status:                True
    Type:                  Completed
```

Reply:
`Processing` is not really a resource change but a condition change

Reply:
I agree the `completed` is not a resource change. but since the common function accepts args which is defined as `ResourceChangeType`, we better not to complicated the situation just want to split the `ResourceChange` and `ConditionChange`, specifically for CDQ ONLY. We have discussed during the team meeting that  for CDQ, we will audit `completed` as a resource change event.

Reply:
updated